pkg/icingaredis/client.go

```diff
@@ -200,6 +204,36 @@ func (c *Client) StreamLastId(ctx context.Context, stream string) (string, error
 	return lastId, nil
 }
 
+// XReadUntilResult (repeatedly) calls XREAD with the specified arguments until a result is returned.
+// Each call blocks at most for the duration specified in Options.BlockTimeout until data
+// is available before it times out and the next call is made.
+// This also means that an already set block timeout is overridden.
+func (c *Client) XReadUntilResult(ctx context.Context, a *redis.XReadArgs) ([]redis.XStream, error) {
+	a.Block = c.Options.BlockTimeout
+
+	XRead := func() *redis.XStreamSliceCmd {
+		// Don't consume go-redis retries if the result is not available in time anyway.
+		ctx, cancelCtx := context.WithTimeout(ctx, a.Block)
+		defer cancelCtx()
+
+		return c.XRead(ctx, a)
+	}
+
+	for {
+		cmd := XRead()
+		streams, err := cmd.Result()
+		if err != nil {
+			if errors.Is(err, redis.Nil) || errors.Is(err, context.DeadlineExceeded) {
+				continue
+			}
+
+			return streams, WrapCmdErr(cmd)
+		}
```

Comment:
Fun fact: you _could_ strip off the outer if (and "just" waste a few CPU cycles).